### PR TITLE
add DETAIL logs for batch use cases

### LINKED ISSUE #10975: Improve logging for batch jobs
_This issue is created to track the thread in the [Lightning slack channel](https://pytorch-lightning.slack.com/archives/CSFA7CXDH/p1636010487113700?thread_ts=1635982958.112600&cid=CSFA7CXDH), summarizing the conversation between @carmocca and @yifuwang_.

## 🚀 Feature

Add additional logging to help debug batch jobs in production.

### Motivation

Currently, Lightning logging optimizes for conciseness for interactive users (`INFO` by default). When using Lightning in production use cases, users may want to have more information to help root cause situations such as: 

* When a job is stuck, experiencing high memory footprint, or just busy spinning somewhere, the log reveals little to no information about what Lightning is doing.
* When a failure happens, we can’t tell from the log what important steps in Lightning took place.
* We can’t tell from the log what implicit decisions Lightning makes based on user input (e.g. a very recent example: Lightning decided to initialize `DistributedDataParallel` even when the user passed `ddp_fully_sharded` as the strategy, causing confusing CUDA OOM issues).
* Failures in batch jobs are very tedious to reproduce. Often the log is the best resource for us to support our users.

`DEBUG` logs may be too verbose, e.g. when we log per iteration.

### Pitch

Create an additional logging level, `DETAIL (15)` between `INFO (20)` and `DEBUG (10)`. 

Add additional logging to critical places to make it easier to determine what decisions were made by Lightning to facilitate debugging. Some example places we could add this to: 

- Trainer
  - All Trainer arguments
  - Start/Train/Test/Predict stages
  - `_run` (prepare data, setup env, restore modules, restore callbacks, etc)
- Loops
  - Advance 
  - On run end
  - Every n batched processed
- Hooks
  - Teardowns
  - Callback hooks (including user code)
- Plugins
  - DDP
    - hook registration info
    - Model shard size
 - Accelerator connector
   - TrainingTypePlugin decisions  

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

Fixes #10975 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -70,6 +74,7 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None) -> None:
         self.resume_checkpoint_path = self.hpc_resume_path or checkpoint_path
         checkpoint_path = self.resume_checkpoint_path
         if not checkpoint_path:
+            log.info("`checkpoint_path` not specified. Skipping checkpoint loading.")
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:77]: should this be verbose?
```suggestion
            log.verbose("`checkpoint_path` not specified. Skipping checkpoint loading.")
```

**Review by tchaton — APPROVED** [review]
Looks great !

File: pytorch_lightning/__init__.py [hunk 2]
```diff
@@ -4,6 +4,18 @@
 
 from pytorch_lightning.__about__ import *  # noqa: F401, F403
 
+VERBOSE = 15  # between logging.INFO and logging.DEBUG, used for logging in production use cases
```
Comment (ananthsub) [inline pytorch_lightning/__init__.py:7]: super nit: verbose as a log level might conflict with `verbose` flags. so if I specify `trainer.verbose_evaluate` it might lead to people thinking logs would come at this level. the issue proposed `DETAIL` which avoids this collision

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 3]
```diff
@@ -100,6 +100,7 @@ def __init__(
             checkpoint_io=checkpoint_io,
             precision_plugin=precision_plugin,
         )
+        log.verbose(f"Initializing DDP: {self.__class__.__name__}")
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:103]: this log line is possibly confusing with L226. 
```suggestion
        log.verbose(f"Initializing DDP Plugin: {self.__class__.__name__}")
```

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 4]
```diff
@@ -329,6 +331,7 @@ def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
         trainer.convert_to_lightning_optimizers()
 
     def configure_ddp(self) -> None:
+        log.verbose(f"{self.__class__.__name__}: configuring DDP...")
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:334]:
```suggestion
        log.verbose(f"{self.__class__.__name__}: configuring DistributedDataParallel...")
```

File: pytorch_lightning/__init__.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/__init__.py:7]:
```suggestion
VERBOSE = 15  # between logging.INFO and logging.DEBUG, used for logging in batch use cases
```

File: pytorch_lightning/__init__.py [hunk 5]
```diff
@@ -4,6 +4,18 @@
...
+
+
+def verbose(self, message, *args, **kws):
+    if self.isEnabledFor(VERBOSE):
+        self._log(VERBOSE, message, args, **kws)
```
Comment (carmocca) [inline pytorch_lightning/__init__.py:12]:
```suggestion
def verbose(self, message, *args, **kwargs):
    if self.isEnabledFor(VERBOSE):
        self._log(VERBOSE, message, *args, **kwargs)
```

File: pytorch_lightning/__init__.py [hunk 6]
```diff
@@ -1,9 +1,22 @@
...
 
+DETAIL = 15  # between logging.INFO and logging.DEBUG, used for logging in production use cases
+
+
+def detail(self: Any, message: str, *args: Any, **kwargs: Any) -> None:
+    if self.isEnabledFor(DETAIL):
+        self._log(DETAIL, message, *args, **kwargs)
+
+
+logging.addLevelName(DETAIL, "DETAIL")
```
Comment (carmocca) [inline pytorch_lightning/__init__.py:16]: You can test this works as intended by using this file

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/utilities/test_warnings.py

where we assert the stdout state. It is run separate from `pytest`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/1fc046cde28684316323693dfca227dbd270663f/tests/standalone_tests.sh#L61-L62

**Review by carmocca — CHANGES_REQUESTED** [review]
Requesting changes for a final review when it's all ready :)

File: pytorch_lightning/__init__.py [hunk 7]
```diff
@@ -1,9 +1,23 @@
 """Root package info."""
 
 import logging
+from typing import Any
 
 from pytorch_lightning.__about__ import *  # noqa: F401, F403
 
+DETAIL = 15  # between logging.INFO and logging.DEBUG, used for logging in production use cases
```
Comment (ananthsub) [inline pytorch_lightning/__init__.py:8]:
```suggestion
_DETAIL = 15  # between logging.INFO and logging.DEBUG, used for logging in production use cases
```

File: pytorch_lightning/__init__.py [hunk 8]
```diff
@@ -1,9 +1,23 @@
...
+
+
+def detail(self: Any, message: str, *args: Any, **kwargs: Any) -> None:
+    if self.isEnabledFor(DETAIL):
+        # logger takes its '*args' as 'args'
+        self._log(DETAIL, message, args, **kwargs)
+
+
+logging.addLevelName(DETAIL, "DETAIL")
```
Comment (ananthsub) [inline pytorch_lightning/__init__.py:17]:
```suggestion
def detail(self: Any, message: str, *args: Any, **kwargs: Any) -> None:
    if self.isEnabledFor(_DETAIL):
        # logger takes its '*args' as 'args'
        self._log(_DETAIL, message, args, **kwargs)


logging.addLevelName(_DETAIL, "DETAIL")
```

File: pytorch_lightning/loops/fit_loop.py [hunk 9]
```diff
@@ -223,6 +224,7 @@ def on_advance_start(self) -> None:  # type: ignore[override]
 
     def advance(self) -> None:  # type: ignore[override]
         """Runs one whole epoch."""
+        log.detail("advancing fit loop")
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:227]: Use the class name

File: pytorch_lightning/loops/fit_loop.py [hunk 10]
```diff
@@ -244,6 +246,7 @@ def on_advance_end(self) -> None:
 
     def on_run_end(self) -> None:
         """Calls the ``on_train_end`` hook."""
+        log.detail("fit loop ended")
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:249]: Use the class name

File: pytorch_lightning/strategies/ddp.py [hunk 11]
```diff
@@ -380,6 +387,7 @@ def pre_backward(self, closure_loss: torch.Tensor) -> None:
             prepare_for_backward(self.model, closure_loss)
 
     def model_to_device(self):
+        log.detail(f"{self.__class__.__qualname__}: moving model to device [{self.root_device}]...")
```
Comment (carmocca) [inline pytorch_lightning/strategies/ddp.py:390]: Why use `__qualname__` here?

Same question for `teardown`

File: pytorch_lightning/trainer/trainer.py [hunk 12]
```diff
@@ -432,6 +432,7 @@ def __init__(
         """
         super().__init__()
         Trainer._log_api_event("init")
+        log.detail(f"{self.__class__.__name__}: Initializing trainer with parameters: {locals()}")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:435]: Nice

File: pytorch_lightning/trainer/trainer.py [hunk 13]
```diff
@@ -744,6 +745,7 @@ def _fit_impl(
         ckpt_path: Optional[str] = None,
     ) -> None:
         Trainer._log_api_event("fit")
+        log.detail("trainer fit stage")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:748]: Use the class name

File: pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -821,6 +823,7 @@ def _validate_impl(
         # SETUP HOOK
         # --------------------
         Trainer._log_api_event("validate")
+        log.detail("trainer validate stage")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:826]: Use the class name

File: pytorch_lightning/trainer/trainer.py [hunk 15]
```diff
@@ -906,6 +909,7 @@ def _test_impl(
         # SETUP HOOK
         # --------------------
         Trainer._log_api_event("test")
+        log.detail("trainer test stage")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:912]: Use the class name

File: pytorch_lightning/trainer/trainer.py [hunk 16]
```diff
@@ -992,6 +996,7 @@ def _predict_impl(
         # SETUP HOOK
         # --------------------
         Trainer._log_api_event("predict")
+        log.detail("trainer predict stage")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:999]: Use the class name

File: pytorch_lightning/trainer/trainer.py [hunk 17]
```diff
@@ -1165,14 +1174,17 @@ def _run(
...
             self._restore_modules_and_callbacks(ckpt_path)
 
         # restore optimizers, etc.
+        log.detail("restoring training state")
         self.checkpoint_connector.restore_training_state()
 
         self.checkpoint_connector.resume_end()
 
         results = self._run_stage()
+        log.detail("tearing down trainer")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1187]: For consistency this should be "__name__ tearing down"

File: tests/utilities/test_warnings.py [hunk 18]
```diff
@@ -54,6 +54,8 @@
     # check that logging is properly configured
     import logging
 
+    from pytorch_lightning import DETAIL
```
Comment (carmocca) [inline tests/utilities/test_warnings.py:57]: You can move this import besides the "block" that uses it (below)

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 19]
```diff
@@ -1165,14 +1174,17 @@ def _run(
         self._log_hyperparams()
 
         if self.strategy.restore_checkpoint_after_setup:
+            log.detail(f"{self.__class__.__name__}: restoring module and callbacks from checkpoint path: {ckpt_path}")
             self._restore_modules_and_callbacks(ckpt_path)
 
         # restore optimizers, etc.
+        log.detail(f"{self.__class__.__name__}: restoring training state")
         self.checkpoint_connector.restore_training_state()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1182]: I disagree with most changes in this PR. Instead of moving the log calls into the actual methods, this is now littering the trainer with more code that is irrelevant to the reader. Our goal is to make the Trainer more readable and easier to trace steps for contributors and researchers who are curious. This PR is going in the wrong direction IMO.
Please reconsider this, I feel very strongly about this. 

@PytorchLightning/core-contributors

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1182]: cc @PyTorchLightning/core-lightning

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1182]: The choice of logging in the caller vs the callee is as old as logging has existed and it is a tradeoff between flexibility vs convenience.

I don't think these log calls make the Trainer unreadable, as they are just glorified prints. You could see their value as a replacement for comments.

> Please reconsider this, I feel very strongly about this.

However, I don't feel very strongly about it, so if you do, I'm okay with changing these.
